tiledb/sm/query/readers/reader_base.cc

```diff
@@ -156,6 +157,59 @@ void ReaderBase::clear_tiles(
   }
 }
 
+bool ReaderBase::need_timestamped_conditions() {
+  // If we have any delete condition that falls between the timestamps of a
+  // fragment with timestamps, generate timestamped query conditions.
+  bool make_timestamped_conditions = false;
+  for (uint64_t i = 0; i < fragment_metadata_.size(); i++) {
+    if (fragment_metadata_[i]->has_timestamps()) {
+      for (auto& delete_condition : delete_conditions_) {
+        auto delete_timestamp = delete_condition.condition_timestamp();
+        auto& frag_timestamps = fragment_metadata_[i]->timestamp_range();
+        if (delete_timestamp >= frag_timestamps.first &&
+            delete_timestamp <= frag_timestamps.second) {
+          make_timestamped_conditions = true;
+          timestamps_needed_for_deletes_[i] = true;
+        }
+      }
+    }
+  }
+
+  return make_timestamped_conditions;
+}
+
+Status ReaderBase::generate_timestamped_conditions() {
```

Comment:
(no change needed now, but:) For testability, we can consider implementing new functions like this as a transforming version (input -> output) , along with a mutating version (this one) which calls the transform. Since the functions inside return `Status`, can use `throw_if_not_ok`.